File: matlab/src/solvers/MVA/solver_mva_retrieval_analyzer.m
function [QN,UN,RN,TN,CN,XN,lG,hitprob,missprob,delayedprob,hitproblist,itemprob,latency,runtime,method] = solver_mva_retrieval_analyzer(sn, options)
% [Q,U,R,T,C,X,LG,HITPROB,MISSPROB,DELAYEDPROB,HITPROBLIST,LATENCY,RUNTIME,METHOD] = SOLVER_MVA_RETRIEVAL_ANALYZER(SN, OPTIONS)
%
% Fixed-point approximation of a delayed-hit (retrieval-system) cache via the FPI
% algorithms: retrieval_fpi (hit / miss / delayed-hit ratios) and
% retrieval_fpi_latency (expected latency Z, eq:latency tot).

% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

T0 = tic;
CN = [];
XN = zeros(1, sn.nclasses);
QN = zeros(sn.nstations, sn.nclasses);
UN = zeros(sn.nstations, sn.nclasses);
RN = zeros(sn.nstations, sn.nclasses);
TN = zeros(sn.nstations, sn.nclasses);
lG = NaN;
method = 'fpi';

line_debug('MVA retrieval analyzer starting: nclasses=%d', sn.nclasses);

[m, lambda, gamma, eta, alpha, T, R, station_type] = cache_retrieval_inputs(sn);
n = numel(lambda);

% --- FPI hit / miss / delayed-hit ratios ---
[pmiss, phit, pdh] = retrieval_fpi(m, lambda, eta, gamma);
pi0  = pmiss(:).';      % per-item miss
pih  = sum(phit, 1);    % per-item hit
phid = sum(pdh, 1);     % per-item delayed

% --- FPI expected latency Z (eq:latency tot) ---
Z = retrieval_fpi_latency(m, lambda, gamma, alpha, T, R, station_type);

% --- read class and per-class aggregates (single read class) ---
ci = find(sn.nodetype == NodeType.Cache);
ch = sn.nodeparam{ci};
rk = keys(ch.retrievalSystemQueueIndices);
jobinClass = double(rk{1}) + 1;
w = lambda(:) / sum(lambda);
hitAgg = sum(w .* pih(:));
missAgg = sum(w .* pi0(:));
delayedAgg = sum(w .* phid(:));

% --- source throughput ---
source_ist = sn.nodeToStation(sn.nodetype == NodeType.Source);
sourceRate = sn.rates(source_ist, :);
sourceRate(isnan(sourceRate)) = 0;
TN(source_ist, :) = sourceRate;

% --- cache results (hitprob is the TRUE hit fraction; delayedprob the
% delayed-hit fraction; true hit + delayed + miss = 1) ---
hitprob     = NaN(1, sn.nclasses);
missprob    = NaN(1, sn.nclasses);
delayedprob = NaN(1, sn.nclasses);
latency     = NaN(1, sn.nclasses);
hitprob(jobinClass)     = hitAgg;
missprob(jobinClass)    = missAgg;
delayedprob(jobinClass) = delayedAgg;
latency(jobinClass)     = Z;

% per-list (per-level) hit fractions for the read class: phit is (h x n);
% access-weighted over items gives the per-list hit probability (rows sum to
% hitAgg over lists).
hitproblist = NaN(sn.nclasses, numel(m));
hitproblist(jobinClass, :) = (phit * w(:)).';

% per-item occupancy [nitems x (lists+1)]: column 1 = miss (item not cached),
% columns 2..end = probability the item resides in each list.
itemprob = [pi0(:), phit.'];

hc = ch.hitclass(jobinClass);
mc = ch.missclass(jobinClass);
if hc > 0, XN(hc) = sourceRate(jobinClass) * (hitAgg + delayedAgg); end
if mc > 0, XN(mc) = sourceRate(jobinClass) * missAgg; end

% --- retrieval-station mean occupancy (QLen) and throughput ---
queueNodes = double(ch.retrievalSystemQueueIndices(rk{1}));
S = numel(queueNodes);
psIdx = find(station_type == "PS" | station_type == "SIRO" | station_type == "FCFS" | station_type == "LCFSPR");   % SIRO/FCFS/LCFSPR as PS
for s = 1:S
    sst = sn.nodeToStation(queueNodes(s));
    if station_type(s) == "PS" || station_type(s) == "SIRO" || station_type(s) == "FCFS" || station_type(s) == "LCFSPR"
        prow = 1 + find(psIdx == s);
    else
        prow = 1;
    end
    phi_s = 0;
    if prow <= size(pdh, 1)
        phi_s = sum(pdh(prow, :));
    end
    tput_s = 0;
    for i = 1:n
        Ri = R(:, :, i);
        a = Ri(1, 2:S+1);
        Pmat = Ri(2:S+1, 2:S+1);
        vis = a / (eye(S) - Pmat);
        tput_s = tput_s + sourceRate(jobinClass) * (lambda(i)/sum(lambda)) * pi0(i) * vis(s);
    end
    QN(sst, jobinClass) = phi_s;
    UN(sst, jobinClass) = phi_s;
    TN(sst, jobinClass) = tput_s;
    if tput_s > 0
        RN(sst, jobinClass) = phi_s / tput_s;
    end
end

runtime = toc(T0);
end

File: jar/src/main/java/jline/solvers/mva/analyzers/Solver_mva_retrieval_analyzer.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.solvers.mva.analyzers;

import jline.api.retrieval.Cache_retrieval_inputs;
import jline.api.retrieval.Retrieval_fpi;
import jline.api.retrieval.Retrieval_fpi_latency;
import jline.lang.NetworkStruct;
import jline.lang.constant.NodeType;
import jline.lang.nodeparam.CacheNodeParam;
import jline.lang.nodes.Cache;
import jline.solvers.SolverOptions;
import jline.solvers.mva.MVAResult;
import jline.solvers.nc.analyzers.Solver_nc_retrieval_analyzer;
import jline.util.matrix.Matrix;

/**
 * Approximate analysis of a delayed-hit (retrieval-system) cache via the FPI
 * algorithms Retrieval_fpi (hit/miss/delayed-hit) and Retrieval_fpi_latency
 * (expected latency Z). Port of matlab/src/solvers/MVA/solver_mva_retrieval_analyzer.m.
 */
public final class Solver_mva_retrieval_analyzer {
    private Solver_mva_retrieval_analyzer() {}

    public static MVAResult solver_mva_retrieval_analyzer(NetworkStruct sn, SolverOptions options) {
        long t0 = System.nanoTime();
        MVAResult res = new MVAResult();
        int K = sn.nclasses;
        Matrix XN = new Matrix(1, K);
        Matrix QN = new Matrix(sn.nnodes, K);
        Matrix UN = new Matrix(sn.nnodes, K);
        Matrix RN = new Matrix(sn.nnodes, K);
        Matrix TN = new Matrix(sn.nnodes, K);
        Matrix CN = new Matrix(1, K);
        Matrix AN = new Matrix(sn.nnodes, K);
        Matrix WN = new Matrix(sn.nnodes, K);

        Cache_retrieval_inputs.Inputs in = Cache_retrieval_inputs.cache_retrieval_inputs(sn);
        int n = in.lambda.length;

        Retrieval_fpi.Result mm = Retrieval_fpi.retrieval_fpi(in.m, in.lambda, in.eta, in.gamma);
        Retrieval_fpi_latency.Result lat = Retrieval_fpi_latency.retrieval_fpi_latency(
                in.m, in.lambda, in.gamma, in.alpha, in.T, in.R, in.stationType);

        double tot = 0; for (double x : in.lambda) tot += x;
        double hitAgg = 0, missAgg = 0, delayedAgg = 0;
        for (int i = 0; i < n; i++) {
            double w = in.lambda[i] / tot;
            double ph = 0; for (double[] row : mm.phit) ph += row[i];
            double pd = 0; for (double[] row : mm.pdh) pd += row[i];
            hitAgg += w * ph; missAgg += w * mm.pmiss[i]; delayedAgg += w * pd;
        }

        Cache cache = null;
        for (int i = 0; i < sn.nodetype.size(); i++) if (sn.nodetype.get(i) == NodeType.Cache) { cache = (Cache) sn.nodes.get(i); break; }
        CacheNodeParam ch = (CacheNodeParam) sn.nodeparam.get(cache);
        Matrix hitProb = new Matrix(1, K), missProb = new Matrix(1, K), latency = new Matrix(1, K);
        Matrix delayedProb = new Matrix(1, K);
        hitProb.fill(Double.NaN); missProb.fill(Double.NaN); latency.fill(Double.NaN); delayedProb.fill(Double.NaN);
        hitProb.set(0, in.jobinClass, hitAgg);
        missProb.set(0, in.jobinClass, missAgg);
        delayedProb.set(0, in.jobinClass, delayedAgg);
        latency.set(0, in.jobinClass, lat.Z);
        // per-list (per-level) hit fractions for the read class: phit is (h x n);
        // access-weighted over items gives the per-list hit probability.
        int h = in.m.length;
        Matrix hitProbList = new Matrix(K, h);
        hitProbList.fill(Double.NaN);
        for (int l = 0; l < h; l++) {
            double acc = 0;
            for (int i = 0; i < n; i++) acc += (in.lambda[i] / tot) * mm.phit[l][i];
            hitProbList.set(in.jobinClass, l, acc);
        }
        // per-item occupancy [n x (h+1)]: column 0 = miss, columns 1.. = per-list.
        Matrix itemProb = new Matrix(n, h + 1);
        for (int i = 0; i < n; i++) {
            itemProb.set(i, 0, mm.pmiss[i]);
            for (int l = 0; l < h; l++) itemProb.set(i, l + 1, mm.phit[l][i]);
        }
        cache.setResultHitProb(hitProb);
        cache.setResultMissProb(missProb);
        cache.setResultDelayedHitProb(delayedProb);
        cache.setResultHitProbList(hitProbList);
        cache.setResultItemProb(itemProb);
        cache.setResultResidT(latency);
        res.hitProb = hitProb; res.missProb = missProb;

        int hc = (int) ch.hitclass.get(in.jobinClass);
        int mc = (int) ch.missclass.get(in.jobinClass);
        if (hc >= 0) XN.set(0, hc, in.sourceRate * (hitAgg + delayedAgg));
        if (mc >= 0) XN.set(0, mc, in.sourceRate * missAgg);

        int sourceNode = -1;
        for (int i = 0; i < sn.nodetype.size(); i++) if (sn.nodetype.get(i) == NodeType.Source) { sourceNode = i; break; }
        // see _kb/06-solver-catalog.md for rationale
        int sourceStation = sourceNode >= 0 ? (int) sn.nodeToStation.get(sourceNode) : -1;
        if (sourceStation >= 0) for (int k = 0; k < K; k++) { double rt = sn.rates.get(sourceStation, k); if (!Double.isNaN(rt)) TN.set(sourceStation, k, rt); }

        Solver_nc_retrieval_analyzer.setStationMetrics(sn, in, mm.pdh, mm.pmiss, QN, UN, RN, TN);

        res.QN = QN; res.UN = UN; res.RN = RN; res.TN = TN; res.CN = CN; res.XN = XN; res.AN = AN; res.WN = WN;
        res.method = "fpi";
        res.runtime = (System.nanoTime() - t0) / 1e9;
        return res;
    }
}
